### ISSUE TITLE
Migration to conda-build 3 and new compilers from Anaconda

### ISSUE THREAD

**Issue by msarahan** [conda-forge/conda-forge.github.io#460]
Mega-thread for discussion and issue tracking with migration.

Currently known things to address:

- [ ] how does toolchain package translate to compiler activation flags with new compilers?
- [ ] Pinning is managed with conda_build_config.yaml files in conda-build 3.  How to migrate conda-forge's existing pinning scheme to that?
- [ ] Compatibility of Anaconda compiler's flags with conda-forge's existing defaults: are the default flags in the anaconda packages good enough that conda-forge need not carry its own customization?  Note: customization risks incompatibility.
- [ ] conda-build-all and job computation.  CB3 has native matrix support, and outputs jobs as rendered MetaData objects.  In Anaconda's CI system, we dump the conda_build_config.yaml files from those rendered objects, and use those dumped conda_build_config.yaml files to build the original templated recipe.  We probably need a way to output temporary folders for the CI's to process as jobs.  CB3's matrix support does not use env vars (but could).
- [ ] Decide on usage of run_exports.  run_exports essentially is the feature where a package declares what its runtime needs are if anyone uses that package at build time.  This feature can make pinning much simpler, but perhaps makes recipes harder to read, because runtime dependencies become indirectly specified.  Anaconda's recipes use run_exports very liberally.

Added 10/18/2017:
- [ ] Implement windows activation script wrappers that are compatible with Appveyor's VS situation

Lots more I'm probably missing here, so please comment and/or modify this issue.

**Comment by isuruf** [comment]
> how does toolchain package translate to compiler activation flags with new compilers?

I think we can remove toolchain. Do the new compilers have flags like `-I${PREFIX}/include` and `-Wl,-rpath,${PREFIX}/lib` ?

**Comment by msarahan** [comment]
I don't think those are defaults. I have needed to add them to some recipes. @mingwandroid, can you tell us why they are not defaults? Are they unsafe in some cases?

**Comment by mingwandroid** [comment]
I would rather not add `-I${PREFIX}/include` because that un-generalizes the compilers.

**Comment by mingwandroid** [comment]
Same for `-Wl,-rpath,...` these things should be added when needed.

**Comment by isuruf** [comment]
> I would rather not add -I${PREFIX}/include or -L${PREFIX}/lib because that un-generalizes the compilers.

I always think of `$PREFIX` as similar to `/usr` where the system compiler `gcc` checks for headers and libraries. In that case it makes sense for the gcc compiler in a conda-build environment to check in `$PREFIX`. This doesn't have to be in a normal conda environment, but can be done in a conda-build environment, just like we disable setuptools downloading in a conda-build environment. (I don't know anything about the new compilers, so I don't have any idea how this can be done)

**Comment by mingwandroid** [comment]
Adding a check for $CONDA_BUILD would be 'ok' in my opinion, still I do not like passing flags to builds that do not need them and serve no purpose.

**Comment by msarahan** [comment]
I would like to add a check for $CONDA_BUILD and add these flags when true.  When builds fail due to missing this, it is an annoying hurdle for people to learn about.  I will try to add this into conda-build today.

**Comment by jakirkham** [comment]
By adding into `conda-build`, do you mean setting environment variables for header and library search paths? If so, strong +1 from me. 😄

**Comment by msarahan** [comment]
In the activate.sh scripts provided by the compiler wrapper packages, gcc_linux-64 & clang_osx-64 

```
if [ "$CONDA_BUILD" == "1" ]; then
    CFLAGS="$CFLAGS -I$PREFIX/include"
    LDFLAGS="$LDFLAGS -L$PREFIX/lib -Wl,-rpath,$PREFIX/lib"
fi
```

CXXFLAGS would be set equivalently in activate scripts for the CXX packages.

(open to suggestions on reordering arguments)

**Comment by msarahan** [comment]
So, I guess to clarify, this isn't going to be set by conda-build, per se.  Conda-build will set CONDA_BUILD, and then the compiler activate scripts will behave a little differently to add these flags.

**Comment by mingwandroid** [comment]
A good example of where `-I${PREFIX}/include` and `-L${PREFIX}/lib` does cause real trouble is when you have a dependency on a package that depends on libiconv on linux. This package should not exist on linux since glibc implements iconv and libiconv was extracted from it.

Adding those flags causes an undeclared dependency to creep in to your build.

So I'm still reluctantly saying 'ok', but it's really wrong IMHO.

**Comment by msarahan** [comment]
Yeah, too bad there's no completely clear win here.  I think the benefits of adding it outweigh the costs, but there also needs to be a way to disable it.

**Comment by mingwandroid** [comment]
Maybe we could delay such a change until conda-build/pyldd is equipped to warn (or better, error) on undeclared dependencies creeping into `DT_NEEDED`?

FWIW, to me it feels like we add those include and link FLAGS to far fewer than half of the recipes in AD5 (but I haven't checked).

**Comment by msarahan** [comment]
Half of the recipes in AD5 is a lot.  If this trips up even 1/100 of the recipes in conda-forge, it's a big waste of people's time.  I'd much rather fix the few recipes that this does affect negatively (these seem to be at the bottom of the stack), and have stuff further up the stack "just work"

**Comment by isuruf** [comment]
There are 198 recipes in `conda-forge` adding ` -L${PREFIX}/lib` which is like 5% of all the recipes or 25% of the recipes with compilation.

**Comment by ocefpaf** [comment]
> A good example of where `-I${PREFIX}/include` and `-L${PREFIX}/lib` does cause real trouble is when you have a dependency on a package that depends on libiconv on linux. This package should not exist on linux since glibc implements iconv and libiconv was extracted from it.

@mingwandroid it does look like the `libiconv` case is a pathological one. (BTW I did not know that, we added `libiconv` everywhere to satisfy some old scientific software and maybe we should revisit its need.)

However, I am more on @isuruf's side. It does looks natural to `conda` do add its `PREFIX` by default.
With that said, I kind of got used to add that manually b/c that is what we've been doing since day one. So I am +0.5.

**Comment by isuruf** [comment]
> CXXFLAGS would be set equivalently in activate scripts for the CXX packages.

At toolchain, we are setting `CPPFLAGS` also

**Comment by isuruf** [comment]
> Pinning is managed with conda_build_config.yaml files in conda-build 3. How to migrate conda-forge's existing pinning scheme to that?

Can conda-build 3 do what `run_exports` is doing using a config.yaml file? If so, it'll be easier to migrate. From that I mean, we pin in the build section and not run where the runtime pinning is based on the buildtime pinning. That way we can update pins in recipes.

Other option would be to remove pinning in recipes and have config.yaml do that. Some issues I have with it are,
1. If there is a dependency that doesn't need to be pinned what happens? For eg: we pin `boost-cpp`, but if a package needs only the boost headers, but not the libraries, what happens?
2. When we update a pinning, right now we look at the individual recipes and update them which is done manually or through a bot. If we switched to a config.yaml, we'll have no idea which recipe needs to be rebuilt or not. (For Anaconda, this is less of an issue because from what I understand, you rebuild everything by looking at dependencies when a pinning change happens, but `conda-forge` is limited by CI)

**Comment by notestaff** [comment]
"-I${PREFIX}/include and -L${PREFIX}/lib does cause real trouble is when you have a dependency on a package that depends on libiconv on linux" -- could the lint checker catch such errors?

**Comment by msarahan** [comment]
For anyone who wants an example of run_exports and centralized pinning, I've been working on libgdal.  here's the recipe: https://github.com/AnacondaRecipes/libgdal-feedstock/blob/master/recipe/meta.yaml

Note that there's only one run requirement.  The built package gets everything else from its build dependencies' run_exports.  The output package metadata looks like:

```
{
  "arch": "x86_64",
  "build": "h28942fd_0",
  "build_number": 0,
  "depends": [
    "curl",
    "expat >=2.2.4,<3.0a0",
    "freexl >=1.0.4,<2.0a0",
    "geos >=3.6.0,<3.6.1.0a0",
    "giflib >=5.1.4,<5.2.0a0",
    "hdf4 >=4.2.13,<4.2.14.0a0",
    "hdf5 >=1.10.1,<1.10.2.0a0",
    "jpeg >=9b,<10a",
    "kealib >=1.4.7,<1.5.0a0",
    "libdap4 >=3.19.0,<3.20.0a0",
    "libkml >=1.3.0,<1.4.0a0",
    "libnetcdf >=4.4.1.1,<4.4.2.0a0",
    "libpng >=1.6.32,<1.7.0a0",
    "libpq >=9.6.5",
    "libspatialite >=4.3.0a,<4.4.0a0",
    "libtiff >=4.0.8,<5.0a0",
    "openjpeg >=2.2.0,<3.0a0",
    "poppler >=0.60.1,<1.0a0",
    "proj4 >=4.9.3,<4.9.4.0a0",
    "sqlite >=3.20.1,<4.0a0",
    "xerces-c >=3.2.0,<3.3.0a0",
    "xz >=5.2.3,<6.0a0",
    "zlib >=1.2.11,<1.3.0a0"
  ],
  "license": "MIT",
  "name": "libgdal",
  "platform": "osx",
  "subdir": "osx-64",
  "timestamp": 1508362884739,
  "version": "2.2.2"
}
```

**Comment by isuruf** [comment]
@msarahan, thanks. That's great. 
I have two questions. 

1. Is it possible to ignore `run_exports`?. For example a python package which loads a dynamic library using `cffi` don't need any pinning.

2. Is it possible to get this output metadata without actually building the package? For the maintenance bot, what we could do is get the output metadata from the latest built package and compare the output metadata if the package was built right now and if not trigger a rebuild.

**Comment by msarahan** [comment]
1. Yes (from https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#pin-downstream):

> The potential downside of this feature is that it takes some control over constraints away from downstream users. If an upstream package has a problematic run_exports constraint, you can ignore it in your recipe by listing the upstream package name in the build/ignore_run_exports section:

```
build:
  ignore_run_exports:
    - libstdc++
```

2. Yes, conda render takes all run_exports information into account.  Here's the outputs from libgdal (note 2 recipes, due to the 2 hdf5 variants):

```
$ conda render libgdal-feedstock
package:
    name: libgdal
    version: 2.2.2
source:
    patches:
        - disable_jpeg12.patch
        - windowshdf5.patch
        - multiprocessor.patch
    sha256: 14c1f78a60f429ad51c08d75cbf49771f1e6b20e7385c6e8379b40e8dfa39544
    url: http://download.osgeo.org/gdal/2.2.2/gdal-2.2.2.tar.gz
build:
    number: '0'
    run_exports:
        - libgdal >=2.2.2,<2.3.0a0
requirements:
    build:
        - glib 2.53.6 ha08cb78_1
        - ld64 274.2 h7c2db76_0
        - sqlite 3.20.1 h900c3b0_1
        - openssl 1.0.2l h6cc03a0_4
        - expat 2.2.4 hf840079_2
        - geos 3.6.0 hd02dd01_1
        - gettext 0.19.8.1 hb0f4f8b_2
        - icu 58.2 hea21ae5_0
        - jpeg 9b haccd157_1
        - hdf5 1.8.18 h1edb405_0
        - openjpeg 2.2.0 hede6aa5_1
        - zlib 1.2.11 h60db283_1
        - postgresql 9.6.5 hf30a1d1_0
        - libspatialite 4.3.0a h3ada6ff_17
        - pcre 8.41 h29eefc5_0
        - freexl 1.0.4 hc15b6b0_3
        - giflib 5.1.4 h6b45fa3_1
        - clang 4.0.1 h662ec87_0
        - xz 5.2.3 h808ee3b_1
        - pixman 0.34.0 h33cc98f_2
        - ca-certificates 2017.08.26 ha1e5d58_0
        - bzip2 1.0.6 h92991f9_1
        - proj4 4.9.3 h8d5df23_6
        - hdf4 4.2.13 h0cfac3f_1
        - curl 7.55.1 h0601bc4_3
        - freetype 2.8 h143eb01_0
        - libdap4 3.19.0 h57690e0_1
        - cctools 895 h7512d6f_0
        - libcxx 4.0.1 h579ed51_0
        - libpng 1.6.32 hce72d48_2
        - poppler 0.60.1 h25db2b2_0
        - libffi 3.2.1 hd939716_3
        - libcxxabi 4.0.1 hebd6815_0
        - cairo 1.14.10 h842a265_5
        - clang_osx-64 4.0.1 hc6fb1d2_0
        - compiler-rt 4.0.1 h5487866_0
        - libxml2 2.9.4 hbd0960b_5
        - libedit 3.1 hb4e282d_0
        - libnetcdf 4.4.1.1 h185d453_7
        - ncurses 6.0 ha932d30_1
        - xerces-c 3.2.0 h4094662_0
        - json-c 0.12.1 h7fa5a77_1
        - llvm 4.0.1 hc748206_0
        - kealib 1.4.7 h20aded0_4
        - libboost 1.65.1 h0af5023_1
        - libkml 1.3.0 he0c3ce7_2
        - libtiff 4.0.8 h8cd0352_9
        - poppler-data 0.4.8 hf2eda46_0
        - libgfortran 3.0.1 h93005f0_2
        - libssh2 1.8.0 h1218725_2
        - fontconfig 2.12.4 h30cff1d_2
        - libiconv 1.15 he69fb81_6
        - llvm-lto-tapi 4.0.1 h6701bc3_0
    host:
        - glib 2.53.6 ha08cb78_1
        - sqlite 3.20.1 h900c3b0_1
        - openssl 1.0.2l h6cc03a0_4
        - expat 2.2.4 hf840079_2
        - geos 3.6.0 hd02dd01_1
        - gettext 0.19.8.1 hb0f4f8b_2
        - icu 58.2 hea21ae5_0
        - jpeg 9b haccd157_1
        - hdf5 1.8.18 h1edb405_0
        - openjpeg 2.2.0 hede6aa5_1
        - zlib 1.2.11 h60db283_1
        - postgresql 9.6.5 hf30a1d1_0
        - libspatialite 4.3.0a h3ada6ff_17
        - pcre 8.41 h29eefc5_0
        - freexl 1.0.4 hc15b6b0_3
        - giflib 5.1.4 h6b45fa3_1
        - xz 5.2.3 h808ee3b_1
        - pixman 0.34.0 h33cc98f_2
        - ca-certificates 2017.08.26 ha1e5d58_0
        - bzip2 1.0.6 h92991f9_1
        - proj4 4.9.3 h8d5df23_6
        - hdf4 4.2.13 h0cfac3f_1
        - curl 7.55.1 h0601bc4_3
        - freetype 2.8 h143eb01_0
        - libdap4 3.19.0 h57690e0_1
        - libcxx 4.0.1 h579ed51_0
        - libpng 1.6.32 hce72d48_2
        - poppler 0.60.1 h25db2b2_0
        - libffi 3.2.1 hd939716_3
        - libcxxabi 4.0.1 hebd6815_0
        - cairo 1.14.10 h842a265_5
        - libxml2 2.9.4 hbd0960b_5
        - libedit 3.1 hb4e282d_0
        - libnetcdf 4.4.1.1 h185d453_7
        - ncurses 6.0 ha932d30_1
        - xerces-c 3.2.0 h4094662_0
        - json-c 0.12.1 h7fa5a77_1
        - kealib 1.4.7 h20aded0_4
        - libboost 1.65.1 h0af5023_1
        - libkml 1.3.0 he0c3ce7_2
        - libtiff 4.0.8 h8cd0352_9
        - poppler-data 0.4.8 hf2eda46_0
        - libgfortran 3.0.1 h93005f0_2
        - libssh2 1.8.0 h1218725_2
        - fontconfig 2.12.4 h30cff1d_2
        - libiconv 1.15 he69fb81_6
    run:
        - giflib >=5.1.4,<5.2.0a0
        - libtiff >=4.0.8,<5.0a0
        - kealib >=1.4.7,<1.5.0a0
        - jpeg >=9b,<10a
        - sqlite >=3.20.1,<4.0a0
        - hdf4 >=4.2.13,<4.2.14.0a0
        - libpng >=1.6.32,<1.7.0a0
        - libdap4 >=3.19.0,<3.20.0a0
        - proj4 >=4.9.3,<4.9.4.0a0
        - poppler >=0.60.1,<1.0a0
        - expat >=2.2.4,<3.0a0
        - curl
        - libpq  >=9.6.5
        - zlib >=1.2.11,<1.3.0a0
        - hdf5 >=1.8.18,<1.8.19.0a0
        - libspatialite >=4.3.0a,<4.4.0a0
        - geos >=3.6.0,<3.6.1.0a0
        - openjpeg >=2.2.0,<3.0a0
        - libnetcdf >=4.4.1.1,<4.4.2.0a0
        - freexl >=1.0.4,<2.0a0
        - xerces-c >=3.2.0,<3.3.0a0
        - libkml >=1.3.0,<1.4.0a0
        - xz >=5.2.3,<6.0a0
test:
    commands:
        - gdal_grid --version
        - gdal_rasterize --version
        - gdal_translate --version
        - gdaladdo --version
        - gdalenhance --version
        - gdalwarp --version
        - gdalinfo --formats
        - gdalinfo http://thredds.nersc.no/thredds/dodsC/greenpath/Model/topaz
        - conda inspect linkages -p $PREFIX libgdal
        - conda inspect objects -p $PREFIX libgdal
    files:
        - test_data
about:
    home: http://www.gdal.org/
    license: MIT
    license_file: LICENSE.TXT
    summary: The Geospatial Data Abstraction Library (GDAL).
extra:
    copy_test_source_files: true
    final: true
    recipe-maintainers:
        - ocefpaf
        - kmuehlbauer
        - gillins
        - msarahan

package:
    name: libgdal
    version: 2.2.2
source:
    patches:
        - disable_jpeg12.patch
        - windowshdf5.patch
        - multiprocessor.patch
    sha256: 14c1f78a60f429ad51c08d75cbf49771f1e6b20e7385c6e8379b40e8dfa39544
    url: http://download.osgeo.org/gdal/2.2.2/gdal-2.2.2.tar.gz
build:
    number: '0'
    run_exports:
        - libgdal >=2.2.2,<2.3.0a0
requirements:
    build:
        - glib 2.53.6 ha08cb78_1
        - ld64 274.2 h7c2db76_0
        - sqlite 3.20.1 h900c3b0_1
        - openssl 1.0.2l h6cc03a0_4
        - expat 2.2.4 hf840079_2
        - geos 3.6.0 hd02dd01_1
        - gettext 0.19.8.1 hb0f4f8b_2
        - icu 58.2 hea21ae5_0
        - jpeg 9b haccd157_1
        - openjpeg 2.2.0 hede6aa5_1
        - libspatialite 4.3.0a h3ada6ff_17
        - zlib 1.2.11 h60db283_1
        - postgresql 9.6.5 hf30a1d1_0
        - pcre 8.41 h29eefc5_0
        - freexl 1.0.4 hc15b6b0_3
        - giflib 5.1.4 h6b45fa3_1
        - clang 4.0.1 h662ec87_0
        - xz 5.2.3 h808ee3b_1
        - pixman 0.34.0 h33cc98f_2
        - ca-certificates 2017.08.26 ha1e5d58_0
        - bzip2 1.0.6 h92991f9_1
        - proj4 4.9.3 h8d5df23_6
        - hdf4 4.2.13 h0cfac3f_1
        - curl 7.55.1 h0601bc4_3
        - freetype 2.8 h143eb01_0
        - libnetcdf 4.4.1.1 h651529c_7
        - libdap4 3.19.0 h57690e0_1
        - cctools 895 h7512d6f_0
        - libcxx 4.0.1 h579ed51_0
        - libpng 1.6.32 hce72d48_2
        - poppler 0.60.1 h25db2b2_0
        - libffi 3.2.1 hd939716_3
        - libcxxabi 4.0.1 hebd6815_0
        - cairo 1.14.10 h842a265_5
        - clang_osx-64 4.0.1 hc6fb1d2_0
        - compiler-rt 4.0.1 h5487866_0
        - libxml2 2.9.4 hbd0960b_5
        - libedit 3.1 hb4e282d_0
        - kealib 1.4.7 h73e85dd_4
        - ncurses 6.0 ha932d30_1
        - xerces-c 3.2.0 h4094662_0
        - json-c 0.12.1 h7fa5a77_1
        - llvm 4.0.1 hc748206_0
        - hdf5 1.10.1 h6090a45_0
        - libboost 1.65.1 h0af5023_1
        - libkml 1.3.0 he0c3ce7_2
        - libtiff 4.0.8 h8cd0352_9
        - poppler-data 0.4.8 hf2eda46_0
        - libgfortran 3.0.1 h93005f0_2
        - libssh2 1.8.0 h1218725_2
        - fontconfig 2.12.4 h30cff1d_2
        - libiconv 1.15 he69fb81_6
        - llvm-lto-tapi 4.0.1 h6701bc3_0
    host:
        - glib 2.53.6 ha08cb78_1
        - sqlite 3.20.1 h900c3b0_1
        - openssl 1.0.2l h6cc03a0_4
        - expat 2.2.4 hf840079_2
        - geos 3.6.0 hd02dd01_1
        - gettext 0.19.8.1 hb0f4f8b_2
        - icu 58.2 hea21ae5_0
        - jpeg 9b haccd157_1
        - openjpeg 2.2.0 hede6aa5_1
        - libspatialite 4.3.0a h3ada6ff_17
        - zlib 1.2.11 h60db283_1
        - postgresql 9.6.5 hf30a1d1_0
        - pcre 8.41 h29eefc5_0
        - freexl 1.0.4 hc15b6b0_3
        - giflib 5.1.4 h6b45fa3_1
        - xz 5.2.3 h808ee3b_1
        - pixman 0.34.0 h33cc98f_2
        - ca-certificates 2017.08.26 ha1e5d58_0
        - bzip2 1.0.6 h92991f9_1
        - proj4 4.9.3 h8d5df23_6
        - hdf4 4.2.13 h0cfac3f_1
        - curl 7.55.1 h0601bc4_3
        - freetype 2.8 h143eb01_0
        - libnetcdf 4.4.1.1 h651529c_7
        - libdap4 3.19.0 h57690e0_1
        - libcxx 4.0.1 h579ed51_0
        - libpng 1.6.32 hce72d48_2
        - poppler 0.60.1 h25db2b2_0
        - libffi 3.2.1 hd939716_3
        - libcxxabi 4.0.1 hebd6815_0
        - cairo 1.14.10 h842a265_5
        - libxml2 2.9.4 hbd0960b_5
        - libedit 3.1 hb4e282d_0
        - kealib 1.4.7 h73e85dd_4
        - ncurses 6.0 ha932d30_1
        - xerces-c 3.2.0 h4094662_0
        - json-c 0.12.1 h7fa5a77_1
        - hdf5 1.10.1 h6090a45_0
        - libboost 1.65.1 h0af5023_1
        - libkml 1.3.0 he0c3ce7_2
        - libtiff 4.0.8 h8cd0352_9
        - poppler-data 0.4.8 hf2eda46_0
        - libgfortran 3.0.1 h93005f0_2
        - libssh2 1.8.0 h1218725_2
        - fontconfig 2.12.4 h30cff1d_2
        - libiconv 1.15 he69fb81_6
    run:
        - giflib >=5.1.4,<5.2.0a0
        - libtiff >=4.0.8,<5.0a0
        - kealib >=1.4.7,<1.5.0a0
        - jpeg >=9b,<10a
        - sqlite >=3.20.1,<4.0a0
        - hdf4 >=4.2.13,<4.2.14.0a0
        - libpng >=1.6.32,<1.7.0a0
        - libdap4 >=3.19.0,<3.20.0a0
        - proj4 >=4.9.3,<4.9.4.0a0
        - poppler >=0.60.1,<1.0a0
        - expat >=2.2.4,<3.0a0
        - hdf5 >=1.10.1,<1.10.2.0a0
        - curl
        - libpq  >=9.6.5
        - zlib >=1.2.11,<1.3.0a0
        - libspatialite >=4.3.0a,<4.4.0a0
        - geos >=3.6.0,<3.6.1.0a0
        - openjpeg >=2.2.0,<3.0a0
        - libnetcdf >=4.4.1.1,<4.4.2.0a0
        - freexl >=1.0.4,<2.0a0
        - xerces-c >=3.2.0,<3.3.0a0
        - libkml >=1.3.0,<1.4.0a0
        - xz >=5.2.3,<6.0a0
test:
    commands:
        - gdal_grid --version
        - gdal_rasterize --version
        - gdal_translate --version
        - gdaladdo --version
        - gdalenhance --version
        - gdalwarp --version
        - gdalinfo --formats
        - gdalinfo http://thredds.nersc.no/thredds/dodsC/greenpath/Model/topaz
        - conda inspect linkages -p $PREFIX libgdal
        - conda inspect objects -p $PREFIX libgdal
    files:
        - test_data
about:
    home: http://www.gdal.org/
    license: MIT
    license_file: LICENSE.TXT
    summary: The Geospatial Data Abstraction Library (GDAL).
extra:
    copy_test_source_files: true
    final: true
    recipe-maintainers:
        - ocefpaf
        - kmuehlbauer
        - gillins
        - msarahan
```

**Comment by msarahan** [comment]
As for build triggering based on stuff already being built, that's actually a weak spot that's on my short-term list.  The current implementation is an exact build string match.  That's way too sensitive and ends up rebuilding when it's not necessary.  It should factor in at least 3 things:

- name
- version
- build variables used

**Comment by jakirkham** [comment]
Like to add on the compiler front that it would be good to sort out how we handle/what we do with the existing LLVM based stack and anything that already relies on them. IOW [`llvmdev`]( https://github.com/conda-forge/llvmdev-feedstock ), [`clangdev`]( https://github.com/conda-forge/clangdev-feedstock ), [`libcxx`]( https://github.com/conda-forge/libcxx-feedstock ), [`openmp`]( https://github.com/conda-forge/openmp-feedstock ), and the pending `flangdev`. Also there is some discussion on adding LLVM's `libunwind` and have the non-GNU `libunwind` already packaged, which may need to be examined in this discussion.

cc @conda-forge/llvmdev @conda-forge/clangdev @conda-forge/libcxx @conda-forge/openmp

**Comment by mingwandroid** [comment]
Why is dev being added to all these package names? I cannot see any reason to do that. You wouldn't name gcc as gccdev for example. What's wrong with flang? Also why was openmp not identified with an llvm- prefix? Names are important and I'd like to know how I can throw my $0.02 in before these things get set in stone.

I have ended up having to create metapackages to point back to these names, e.g.:

https://github.com/AnacondaRecipes/boost-cpp-feedstock/blob/master/recipe/meta.yaml#L139-L159

With openmp, we might have wanted to use that name as a variant selecting metapackage but can't now.

**Comment by msarahan** [comment]
As Ray says, I recommend creating metapackages for these.  They would point to compiler activation packages (e.g. clang_osx-64), which in turn point to "meat" compiler packages - the actual executables and all that.

The more we can agree on flags and centralize compiler packages (not duplicate packages on both conda-forge and defaults), the better (for the sake of cross-channel compatibility).  I understand the arguments for keeping conda-forge self-contained, but I do feel like some compromise may be helpful here. We should discuss this in a call, though.  Comments aren't really adequate.

What Ray is talking about regarding openmp is that there is more than one implementation.  Intel also has one.  Over time, it will be wise to ditch the existing meaning of the ``openmp`` package name, and use that name instead as a metapackage that can select different implementations.

**Comment by SylvainCorlay** [comment]
> Names are important and I'd like to know how I can throw my $0.02 in before these things get set in stone.

The Debian community has very strict naming conventions for packages. For example, python packages are generally named `python-pypi_package_name` which prevents conflicts with ruby, perl and whatnot.

As conda becomes used more and more as a general-purpose package manager, instead of being focused on python, a conda-based distribution would benefit a lot from using such naming conventions.

**Comment by msarahan** [comment]
I doubt you'll find any dissent to that, but we also have to have a
comprehensive plan for migrating from old names to new ones. The
metapackages Ray and I have talked about are how we think it needs to be
done. The other question is the timescale for declaring old names invalid.
While perpetual maintenance may seem friendly to users, it is burdensome
and also precludes using old names for new purposes like openmp here.

On Nov 2, 2017 7:43 AM, "Sylvain Corlay" <notifications@github.com> wrote:

Names are important and I'd like to know how I can throw my $0.02 in before
these things get set in stone.

The Debian community has very strict naming conventions for packages. For
example, python packages are generally named python-pypi_package_name which
prevents conflicts with ruby, perl and whatnot.

As conda becomes used more and more as a general-purpose package manager,
instead of being focused on python, a conda-based distribution would
benefit a lot from using such naming conventions.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/conda-forge/conda-forge.github.io/issues/460#issuecomment-341409428>,
or mute the thread

**Comment by msarahan** [comment]
I wrote up a CEP for the renaming issue at https://github.com/conda/conda/wiki/CEP-2:-Renaming-packages.  Feedback welcome.

**Comment by SylvainCorlay** [comment]
Thanks, that would be quite awesome!

**Comment by ChrisBarker-NOAA** [comment]
All good, thanks folks!

But frankly, we've been really lax about using proper conventions for naming new packages too!

for instance, as a rule, new python packages in conda forge generally simply get their PyPi name, Given conda's legacy, maybe should keep doing that, but there should be a clear policy.

**Comment by jakirkham** [comment]
So given the big topics this issue is trying to tackle, I'd like us to avoid digressing into a general discussion about naming. Not to say that isn't a big topic worthy of attention. Just there is a long history behind that one already and it really deserves its own thread. Maybe issue ( https://github.com/conda-forge/conda-forge.github.io/issues/18 ) or a new one continuing off of that issue. Could be called "Naming revisited". To be clear happy to have that discussion continue in tandem. Just not in this thread. Thanks in advance for understanding.

**Comment by jakirkham** [comment]
That said, the naming of the compiler packages mentioned is definitely relevant for this thread and should be discussed. Responding to the points raised by @mingwandroid on this front below. Sorry if I missed anything. Much to say/respond to on this front. Tried to respond per package. Then some musings on metapackages and more generally where we stand on the compiler front at the end.

As to your question about `llvmdev`, here is the original PR ( https://github.com/conda-forge/staged-recipes/pull/1482 ) that added it. Basically a year old. We can all probably benefit from skimming it at least. The name switched several times. I believe the dev was added because it builds static libraries and version was left out of the name as the libraries are static. I think the same is true of `clangdev`, which was added in PR ( https://github.com/conda-forge/staged-recipes/pull/1481 ). Though @inducer will remember better than I.

We added `openmp` in PR ( https://github.com/conda-forge/staged-recipes/pull/3171 ). We debated `openmp` (Homebrew uses this) vs. `llvm-openmp` (didn't find anyone using this, but it must be out there) vs. `libomp` (Linux distros use this) in this [thread]( https://github.com/conda-forge/staged-recipes/pull/3171#pullrequestreview-47032873 ). In the end we went with `openmp` as we saw no conflict with `gcc` as it would likely use `libgomp`. There wasn't much concern with an `intel` conflict at the time even though I raised it. Probably as it was unclear if Intel packages would exist in conda-forge (though this is speculation as to the lack of interest from others). There has been some interest in changing this both on the conda-forge and Intel ends, but that is a totally different discussion that should be separate ( https://github.com/conda-forge/staged-recipes/issues/3022 ) and probably needs a meeting as well. Digression aside we stuck with `openmp` as it matched the package name and was the most clear to people. Not attached to it though, but renaming would break a few things now (not many though).

`flangdev` is still in the works. Guessing the name also relates to static libraries as well. Though I have not been close to that work. My gut feeling is much of that is still in flux and guessing the name could be changed. There is not a staged-recipes PR for it yet. Would have to talk with @isuruf for details.

The metapackage idea is interesting. Particularly with `openmp`. Though it feels like features are still pretty broken. What the status is of the build string in `conda-build` 3. That said, tweaking the build string still feels like a hack (admittedly better than nothing). Is there any other work on this front in `conda`/`conda-build` that would improve package selection.

Though I guess the question I have right now is are we only concerned about the names of these compilers. If so, that makes me feel a lot better about where we stand, but I feel this is false confidence. We probably need to take a look at how LLVM and friends are built in both channels before we can really assess what if anything needs to change. I hesitate to spin this off as a subthread/new issue yet, but maybe we should be thinking about that too.

**Comment by mingwandroid** [comment]
OK, let's save naming for another time, but I will say adding `dev` just because there are static libs seems silly to me. Lots of things have static libs. The most important thing about names apart from avoiding conflicts, is not violating the principle of least surprise.

Anyway ...

The clang recipe is available here: https://github.com/AnacondaRecipes/aggregate/tree/master/llvm-compilers-feedstock/recipe

It is huge and complicated. It generates separate packages for `llvm-lto-tapi `, `cctools`, `ld64`, `clang`, `clangxx`, `llvm`, `polly`, `libcxx`, `libcxxabi `, `libunwind`, `llvm-openmp` and `compiler-rt`.

The complexity was necessary due to needing to build on macOS 10.9 but to use a more recent clang than macOS 10.9 comes with to be able to build clang 4.0.1 - we use an official clang binary here - and while those clang binaries do run on 10.9 they need a more recent version of `cctools` (specifically `libtool` and `ranlib`) than macOS 10.9 provides. Building `cctools` and `ld64` requires two parts of `llvm` (`libLTO` and `libtapi`) so we must build bits of `llvm` then build `cctools` then build the rest of `llvm` and `clang`. Then after that, I re-build *everything* again to ensure that *nothing* has leaked in from the system tools nor from the original official 4.0.1 clang binaries.

So overall, I would recommend we take a different approach. I propose to make this recipe build a set of packages suffixed with `-bootstrap`. No one need ever worry about how to build those (unless they want to join my pain). It does build fine and is repeatable (I must have built these about 20 times during development), but there is so much stuff there that making changes is always a bit of a minefield.

I don't know if people care about having cleaner recipes to rebuild `cctools` and `ld64` packages (Apple's `binutils` basically), but if so then we need a recipe for `llvm-lto-tapi` that depends on the `-bootstrap` toolchain and a single split-package recipe for `cctools` and `ld64` that depends on `llvm-lto-tapi` and the `-bootstrap` toolchain, then a second recipe for all of the rest (including anything that can be built in-tree with the latest llvm code, so `Julia` is probably out but `flang` and `rust` are probably fine).

Doing it this way means we can free this new llvm split-packages recipe from the hoops we need to jump through to modernize a macOS 10.9 toolchain and from the Apple specific `cctools` and `ld64` bits which should make it much easier for anyone familiar with 'normal' llvm to hack on it. If the bootstrap compilers ever need to be updated then I'll take that on.

**Comment by msarahan** [comment]
with https://github.com/conda-tools/conda-build-all/pull/93 and conda-build 3.0.28, conda-build-all is now compatible with conda-build 3.

**Comment by mingwandroid** [comment]
Before I spend time on this I need to know if it's an acceptable way forward.

Specifically, the `conda-forge` `llvm`, `clang`, `cctools` and the `ld64` `build.sh` would need to install some new 'bootstrap' compilers (which are just renames of these ones) into a temporary location, add that to PATH and manually run the activation scripts then the outputs from building these recipes would become the compilers we all use for macOS.

I would also like to see consensus on how the recipes are split into packages (i.e. what goes into the sub/split-packages, from my perspective a single recipe should build as much as possible that's cross-platform or intends to become cross-platform).

**Comment by isuruf** [comment]
Why not use the compilers from defaults? If we try to build everything in one recipe, CI won't be able to build them.

**Comment by mingwandroid** [comment]
> Why not use the compilers from defaults?

I do want to use them but to use them to build clang and llvm and the rest this needs special attention (manual installation and sourcing of activate scripts). The compilers are inextricably interwoven with the `llvmdev` and `clangdev` packages. They cannot co-exist and cannot be built with one depending upon the other without using the `always_include_files` hack and that would be horribly brittle.

> If we try to build everything in one recipe, CI won't be able to build them

The overhead of adding things to an llvm recipe is pretty low since the majority of build time is for llvm itself.

Even if it does exceed CI limits I do not think that should be allowed to limit the quality of how we construct these recipes. We make exceptions to upload binaries for other packages that have long builds and I believe the quality of the llvm recipes is very important.

**Comment by isuruf** [comment]
> I do want to use them but to use them to build clang and llvm and the rest this needs special attention (manual installation and sourcing of activate scripts). 

What I'm suggesting is to use the defaults compilers if we want to compile something, but use the conda-forge packages if we want `llvm` to be linked to a package which would need to JIT compile a function like `llvmlite` and `pocl` which needs access to `libclang`

> The compilers are inextricably interwoven with the llvmdev and clangdev packages. They cannot co-exist and cannot be built with one depending upon the other without using the always_include_files hack and that would be horribly brittle.

Okay. So is this a macos only issue where the compiler is also clang? Can the compilers be installed somewhere other than `PREFIX`, so that we can build llvm?

> The overhead of adding things to an llvm recipe is pretty low since the majority of build time is for llvm itself.

building clang takes a lot of time as well.

> Even if it does exceed CI limits I do not think that should be allowed to limit the quality of how we construct these recipes. We make exceptions to upload binaries for other packages that have long builds and I believe the quality of the llvm recipes is very important.

One other issue is that if we wanted a different variant of one package, what do we do? For example, `flang` needs patches for `clang` and `cling` needs patches for `llvm`

**Comment by mingwandroid** [comment]
> What I'm suggesting is to use the defaults compilers if we want to compile something, but use the conda-forge packages if we want llvm to be linked to a package which would need to JIT compile a function like llvmlite and pocl which needs access to libclang

The 'but' is the issue, all of the time this is an 'and'.

> Okay. So is this a macos only issue where the compiler is also clang?

The compiler is *always* clang on macOS and I want to enable it on Linux as soon as possible too (though not as the default). Then I want to investigate using the clang GCC frontend with the clang MSVC backend on Windows to allow us to build things that use the GNU stack with Windows UCRT support (probably in a year or two!).

> Can the compilers be installed somewhere other than PREFIX, so that we can build llvm?

By this do you mean that the compilers are built to install to something like `${PREFIX}/clang` so that is added to `PATH` whenever something needs to use these compilers? If so I'd rather not, and I have explained what I think is a better approach in [some detail](https://github.com/conda-forge/conda-forge.github.io/issues/460#issuecomment-341517704).

> One other issue is that if we wanted a different variant of one package, what do we do? For example, flang needs patches for clang and cling needs patches for llvm

Presumably the patches are intended to end up upstream? If so they *should* not break other things, therefore you might be able to use the superset. Please point me at the patches and I will investigate as best I can.

**Comment by isuruf** [comment]
> The 'but' is the issue, all of the time this is an 'and'.

I'm not sure what you mean by this.

> By this do you mean that the compilers are built to install to something like ${PREFIX}/clang so that is added to PATH whenever something needs to use these compilers? If so I'd rather not, and I have explained what I think is a better approach in some detail.

Can you explain your reasons as to why this is not viable?

> Presumably the patches are intended to end up upstream? If so they should not break other things, therefore you might be able to use the superset. Please point me at the patches and I will investigate as best I can.

See https://github.com/flang-compiler/clang and https://github.com/flang-compiler/clang/pull/38 for the windows patches on top of that. (These are for 4.0 branch, there are branches and PRs for 5.0 branch as well)

Btw, what's your suggestion for linux and windows?

**Comment by msarahan** [comment]
> The 'but' is the issue, all of the time this is an 'and'.
> I'm not sure what you mean by this.

I think what Ray means is that he wants people to use the same compiler all of the time.  He thinks that we need to use our compiler that we have right now as a bootstrapping compiler to build the CF recipe, and then use the CF recipe to build everything.  There is no "use this compiler for A, and this compiler for B" - only how you bootstrap one compiler.  That one compiler is the CF recipe, and the bootstrap compiler is our new one.

Usually, setting something like $PREFIX/bin/clang is a bad idea because clang is coded to detect the target platform from the clang executable filename.  Keeping this lined up for cross compiling is important, I think.  @mingwandroid may have more to add.

For Windows, I think we're sticking with MSVC activation scripts.  Anaconda will continue to use the Intel fortran compiler for now.  I think we could consider flang replacing it once it's proven.

For linux, is there any reason not to use Anaconda's new gcc packages?  That doesn't encroach on any existing work in CF, so I don't see any conflict here.  It's just replacing devtoolset2 with the new stuff.

**Comment by isuruf** [comment]
For windows and linux, I was asking about llvm and clang.

**Comment by mingwandroid** [comment]
Seems `flang` is nothing but a divergent fork of `clang`? Urgh, shame. OK, we should ignore `flang` (from the perspective of building it at the same time as the rest of the llvm ecosystem) until they have upstreamed their work, which will probably happen once they get sick of rebasing.

**Comment by mingwandroid** [comment]
> Btw, what's your suggestion for linux and windows?
> For windows and linux, I was asking about llvm and clang.

I don't know what you mean, nothing special needs to be done for Windows or Linux since they do not use clang as the default system compilers. Please explain.

**Comment by isuruf** [comment]
> I don't know what you mean, nothing special needs to be done for Windows or Linux since they do not use clang as the default system compilers. Please explain.

I mean, what happens to the recipe? Is that going to be the same as in Mac OS X?

> OK, we should ignore flang (from the perspective of building it at the same time as the rest of the llvm ecosystem) until they have upstreamed their work, which will probably happen once they get sick of rebasing.

You mean don't package flang at all in conda-forge?

**Comment by msarahan** [comment]
If it's already packaged, why remove it?  I think the tricky question is in maintaining binary compatibility, and that's all we're concerned about here.  We don't want to take away useful tools.  We do want to make software between defaults and CF compatible.

So, if flang is used on windows for odd things, fine.  If flang is used on windows for all fortran code (numpy, blas, etc), then we need to figure out how to unify.

**Comment by isuruf** [comment]
> If flang is used on windows for all fortran code (numpy, blas, etc), then we need to figure out how to unify.

That's the goal. flang can compile the reference-lapack without any issue and all tests pass. OpenBLAS can be compiled (with assembly routines and fortran files). SciPy can be compiled. (3 tests from the full test-suite fail though)

**Comment by mingwandroid** [comment]
I've asked this before. How do you hope to use flang on Windows when it requires using the UCRT? What do you do about numpy for Python 2.7? Do you wait until the end of Python 2.7?

**Comment by mingwandroid** [comment]
But anyway, no I was not suggesting removing flang, only removing it from this discussion about what parts of the llvm ecosystem to build all at once and how to migrate to new compilers since it does not (as things stand) have a dog in that race (due to being a complete fork of clang).

**Comment by isuruf** [comment]
> I've asked this before. How do you hope to use flang on Windows when it requires using the UCRT? What do you do about numpy for Python 2.7? Do you wait until the end of Python 2.7?

I have no plans for that. Use defaults is the best option. Having python 3.5+ support is better than no support at all.

> due to being a complete fork of clang

Yes, but the changes they've made are minimal and not affecting C/C++ stuff. See https://reviews.llvm.org/D34097 for a diff of the changes.

**Comment by msarahan** [comment]
You seem to have missed my point about unification.  How will your new flang-built packages interact with defaults?  That is, packages built with gfortran on mac, new gfortran on linux, and ifort on windows?

I would posit that introducing something that breaks compatibility with defaults is worse for the community than not having it at all and depending on defaults.

The last thing I want is to keep conda-forge from building fortran code, but I also really must insist that CF tries to gain that ability in a way that maintains compatibility with defaults.  We (anaconda) get the brunt of support requests from newbies, and incompatibility with defaults places an undue burden on us.  This issue, and our compiler effort as a whole, is all about getting the two channels to coexist better.  Let's please focus on that.

**Comment by msarahan** [comment]
PS: I'm not opposed to pushing for defaults to switch to flang, either, but that is a different topic.  There, it must be weighed against the existing implementation (gfortran or ifort), and we'd have to evaluate any performance differences.  Flang would need to be well-proven and accepted as an official llvm project, too, I think.

**Comment by mingwandroid** [comment]
> I have no plans for that. Use defaults is the best option. Having python 3.5+ support is better than no support at all.

It really saddens me that you think this is acceptable, either for users of defaults or users of conda-forge.

**Comment by msarahan** [comment]
Let's have a meeting for this.  I've created a whenisgood poll for times next week: http://whenisgood.net/7ihn88f

For visibility, I'll email core team members directly.  Everyone is welcome, of course.  I can host the call on webex if we exceed the 10-person limit on hangouts.

**Comment by msarahan** [comment]
The magic time looks like Friday, Nov. 17, 9 AM Central time.  See you all soon.

**Comment by jakirkham** [comment]
Sorry @msarahan. I haven't even had a chance to look at this poll or my schedule (the next few days are a bit insane) and imagine there are a few others in the same boat. Considering this just got posted today and the importance of the compiler discussion to so many. Could we please hold off on declaring a date for a day or two so people have time to look?

**Comment by msarahan** [comment]
I don't really see what waiting another day or two will buy:

<img width="458" alt="screen shot 2017-11-08 at 13 52 32" src="https://user-images.githubusercontent.com/38393/32571175-2804ef5e-c48c-11e7-85bd-9adccb9c89fa.png">

That includes responses from most of the core team:

Jonathan Helmus
Filipe Fernandes
Matt Craig
Isuru
Phil Elson
Ray Donnelly
Eric Dill
Sylvain Corlay

Realistically, we will probably not be able to fit absolutely everyone into a schedule.  I can put off "officially" scheduling it, but I'm marking Nov 17 at 9 AM on my calendar.

**Comment by jakirkham** [comment]
I don't understand the objection @msarahan. Is there any harm from waiting a day or two longer for feedback on the date?  As we have generally done the same with general conda-forge meetings, I think it is reasonable to do the same with this one.

Would add that presumably part of the reason for this meeting isn't just to discuss the compiler proposal and state of affairs, but to agree on a plan and get buy-in, correct? Waiting another day or two for people to respond on the date only helps show that an earnest effort was put forth on making it accessible. Racing the scheduling along however does not (even if the only goal is to ensure we have a date on the books ASAP). In light of this would encouraging waiting a day or two.

**Comment by jakirkham** [comment]
Also would like to reach out @wesm and @inducer to make sure they are aware of this meeting as I know both have a fair bit of interest in compiler tooling from past work.

**Comment by msarahan** [comment]
OK.  Let's make the final call by this Friday, Nov 10.  I will open the poll to extend to the following week, as well.

**Comment by wesm** [comment]
I don't have strong views other than moving to devtoolset-3 (or higher) so that C++14 features are available should they be needed

**Comment by jakirkham** [comment]
Thanks @msarahan.

**Comment by jakirkham** [comment]
So the discussion is about leaving system compilers behind entirely and using Anaconda's and/or conda-forge's built compilers in some fashion. Admittedly I'm doing a hack job summarizing that. So please don't read too much into that. There are some exceptions like Windows. Though the plan is devtoolset will be gone. Hope that is a bit clearer, @wesm.

**Comment by jakirkham** [comment]
Would be great if you could come to this meeting as well, @xoviat.

**Comment by msarahan** [comment]
Anaconda does ship MinGW binaries.  @mingwandroid is a major msys2 contributor.  They might be a bit out of date now, but we have much of msys2 as conda packages.

Why is gfortran with msys2 not an option?  Why is gfortran on Mac not an option?  

We don't need to debate this here.  It can wait for the meeting.

**Comment by xoviat** [comment]
> We don't need to debate this here. It can wait for the meeting.

Unfortunately, I won't be able to attend the meeting. But what I will say is that this is not a debate that is happening here. Numpy *will* drop support and then everyone else will have to deal with it (that is unless people who don't bear the maintenance cost object to the plans).

> Why is gfortran with msys2 not an option?

Scipy and Numpy have to ship CRT binaries to comply with the Python specification. This is not an issue that is up for debate, except on the python-dev mailing list. These projects *will* drop MSVC 2008 as the need to upgrade to C++11 becomes increasingly important. And that means dropping Python 2.7.

If you don't like this, then it's too bad. Because this is the future that python-dev has designed, and the decision goes all the way up to the BDFL. It's probably too late to undo at this point.

**Comment by ChrisBarker-NOAA** [comment]
OT, but briefly:

On Thu, Nov 9, 2017 at 1:43 PM, xoviat <notifications@github.com> wrote:

> Scipy and Numpy have to ship CRT binaries to comply with the Python
> specification. This is not an issue that is up for debate, except on the
> python-dev mailing list.
>
> sure it is -- the python.org build of 2.7 for Windows is built with MSVC
2008 -- and that is not going to change.

But IF (and this is a big IF) there is a community demand for py2.7 with a
newer compiler (either before or after 2.7 officially reaches end of life),
then there is no technical reason Anaconda.com or conda-forge, (or whoever)
couldn't provide a  python2.7 built with a new and better compiler.

That hasn't been done so far, as compatibility with the python.org builds
has been considered important. But at some point, being able to build
extensions, etc with a modern compiler may be compelling.

However: there are a LOT of reasons to drop 2.7 support anyway -- getting a
newer compiler on Windows should be the least of the reasons. So it looks
like numpy is going to declare a drop date -- and we can all have one less
"platform" to support, so I'll be happy for that day to come

(even though I haven't ported my major projects yet!)

-CHB

**Comment by isuruf** [comment]
> Anaconda does ship MinGW binaries. @mingwandroid is a major msys2 contributor. They might be a bit out of date now, but we have much of msys2 as conda packages.

Yes, but Anaconda keeps them separate, and use them rarely (like rpy2) when building a python extension. 

> Why is gfortran with msys2 not an option? Why is gfortran on Mac not an option?

gfortran on Mac is the best option and flang should be considered only for windows

**Comment by xoviat** [comment]
@ChrisBarker-NOAA 

I'm sure you already know, but I qualified my statement with "except on the python-dev mailing list." So essentially, what would have to happen is someone would have to start a thread to release a Python 2.8 with a newer ABI. Because projects are not going to use a Python that is not from Python.org.

Yes, theoretically it is up for debate (as I qualified). But not here. It's up for debate on python-dev. And it's up for debate in a very hostile environment that's (justifiably) decided on rules (PEP 408) that prohibit  exactly the kinds of changes that would need to happen. I'll tip my hat to anyone who manages to get through that.

**Comment by xoviat** [comment]
> Why is gfortran on Mac not an option?

Also, I should point out that gfortran on mac *is* an option precisely because it *is* compatible with the native (clang) ABI. If gfortran on windows was compatible with the native (UCRT) ABI, then it would be an option on windows as well. But it's not, so it isn't.

**Comment by ChrisBarker-NOAA** [comment]
@xoviat:

Actually, that was my point -- python-dev has nothing to do with it -- if  conda-forge or anaconda.org want to release a 2.7 built with a different compiler on Windows, they can do so -- it just won't be compatible with the python,org build.

This isn't all that different than the LInux situation -- plenty of binaries of pythonx.y that aren't binary compatible with each-other (at least before many linux...)

The community came to a consensus years ago that standardizing eh Windows build was the best way to go -- but times have changed -- MAYBE it's time for a new approach (NOTE: I"m not advocating for it -- I think there are other reasons to go to py3, so non need to stretch out support for py2)

as for "release a Python 2.8 with a newer ABI." - That was proposed a couple years ago by Christian Tismer for Stackless. It was soundly rejected.

Part of the point was that a 2.7 built with a different ABI is still 2.7, just like 2.7 on Mac, or Linux, or . is still 2.7. A new ABI is essentially a new platform, not a new python version.

And indeed, there is a (at least source) for a VS2010 build of 2.7:

https://github.com/ctismer/Python-2.7.5-VS2010

And as for: "projects are not going to use a Python that is not from Python.org. -- not so sure about that, plenty of people are building stuff on top of Anaconda/conda -- it's compatible with python.org, but it's not "from" python.org.

This could all be done -- but it would require a significant effort -- and I think that effort should be better spent on porting ot py3 -- then you get py3 and newer compilers for free!

(I still have no idea why Stackless hasn't gone to py3 -- wait! maybe it has:

https://github.com/stackless-dev/stackless

Which might be why Cristian gave up on "2.8"

**Comment by xoviat** [comment]
> Actually, that was my point -- python-dev has nothing to do with it -- if conda-forge or anaconda.org want to release a 2.7 built with a different compiler on Windows, they can do so -- it just won't be compatible with the python,org build.

Yes, but numpy and scipy will just ignore this build and drop support anyway. That's what I'm saying.

**Comment by xoviat** [comment]
The moment that projects need to use C++11, they will drop support. The moment they need to use fortran, they will drop support. And then the cleanups begin. Removing all of the Python 2.7 hacks because "we don't need them anymore." And reworking the C API to take advantage of newer Python improvements. And then no matter what you do, you cannot build these projects on Python 2.7.

**Comment by mingwandroid** [comment]
While I like where flang is going, realistically it is probably 2 years away from being usable by any serious software distribution as their fortran compiler.

While I'm here I would like to clear up some misconceptions that have been thrown about regarding Microsoft C runtime compatibility.

Yes, mingw-w64 links to msvcrt.dll. You can load a DLL (such as a python module) that links to that into a Python that's linked to the vs2008 CRT or to the UCRT just fine. The only problem would be if you allocated something from a DLL linked to one C runtime and deallocated it in another or passed OS-level primitives between them. In reality this basically never happens as it would be very poor software design. I have certainly not see it.

This is not some theoretical musings, this is reasonably large part of my job. I have built a large amount of python modules that link to mingw-w64 compiled DLLs. Most recently libopus and libvpx linked to OpenCV, running on Python 2 and 3.

Also Theano dynamically compiles kernels as python modules using mingw-w64 and loads them dynamically.

It is entirely possible (and very easy) to use mingw-w64 with Cython.

Our RStudio does something very similar with a program called rsession.exe. It is compiled with VS2015 linking to the UCRT but the vast majority of what it does is done via the r-base dlls and those are all compiled with mingw-w64.

Relatedly, all of conda-forge's (and default's) R infrastructure on Windows is compiled with mingw-w64.

So I hope @xoviat has not managed to fool *anyone* into thinking this linking is impossible (or even highly problematic). Blatantly it is not.

To get on topic, most fortran software expects to be compiled with either gfortran or something that emulates gfortran's CLI. It is *usually* driven through Makefiles or autotools. Once flang is mainlined into llvm/clang and once it's been battle hardened by a few years of 'proper' use and once it emulates gfortran's CLI on Windows and once Python 2.7 has actually died, I will happily consider recommending it instead of gfortran.

Until then you flang pushers are just dreaming.

**Comment by xoviat** [comment]
I agree with that; I even noted: 

>  The only advantage for flang is that it's open source, and thus has a compatible license. Otherwise, it basically for now does almost everything that Intel does but performs worse.

My central claim was that projects will drop support for Python 2.7.

> So I hope @xoviat has not managed to fool anyone into thinking this linking is impossible (or even highly problematic). Blatantly it is not.

Not only do I know this, and not only have I enabled SciPy to be build on Python 2.7 with this knowledge, but I have been quite devious. I have convinced a number of projects not to release Python 2.7 wheels on pypi.org for windows in order to inflict maximum pain on Python 2.7 windows users. Coincidentally, this has saved me a not-significant amount of effort in packaging these projects.

> Until then you flang pushers are just dreaming.

I'll refer to my earlier statement. I don't disagree that flang is not the ideal compiler. But it gets the job done for many projects, which are usually written in F77 or F90 (flang has some problems with F2008).

> The future that I see is that everything is basically compiled with MSVC/flang.

**Comment by mingwandroid** [comment]
> Not only do I know this, and not only have I enabled SciPy to be build on Python 2.7 with this knowledge

Please show evidence of this claim. I backed up the claims I made with copious amounts of this, so far you have provided none.

**Comment by xoviat** [comment]
For reference, the `vtk` package only has Python 3.6. `scikit-umfpack` only has Python 3.5 and Python 3.6. And there are even more projects similar to this.

**Comment by xoviat** [comment]
I mean come on: https://github.com/scipy/scipy/pull/7616

**Comment by mingwandroid** [comment]
> For reference, the vtk package only has Python 3.6. scikit-umfpack only has Python 3.5 and Python 3.6. And there are even more projects similar to this.

.. because they use C++11 right? Or is it because you've convinced them not to bother? If it's either of those then that's got nothing to do with the issue of fortran though and I would ask that you please stop conflating them, credit your audience with some intelligence.

That you are taking and admitting to these acts of sabotage against Python 2.7 on Windows should rule you out from further involvement in this discussion as far as I am concerned. You are far from objective.

FWIW, I do not enjoy Python 2.7 on Windows, but I am not going to let that blind me to the correct course of action.

**Comment by xoviat** [comment]
> Or is it because you've convinced them not to bother?

No. It's because I chose not to package Python 2.7.

> If it's either of those then that's got nothing to do with the issue of fortran though and I would ask that you please stop conflating them, credit your audience with some intelligence.

No, you're right that it has nothing to do with fortran. Again, I reiterate:

> My central claim was that projects will drop support for Python 2.7.

> That you are taking and admitting to these acts of sabotage against Python 2.7 on Windows should rule you out from further involvement in this discussion as far as I am concerned. You are far from objective.

You're correct. I am not objective. But I only have the same view as most of the core developers (@ChrisBarker-NOAA has the same view but just phrases it more nicely). And you might consider why I have this view: every single time I write code, it's always Python 2.7 that fails on the CI. Always. I'm sure that I'm not alone. And I'm biased precisely because I've had to spend extra time supporting a language that I've never used. My only goal is to make the development process seamless and Python 2.7 gets in the way of that.

**Comment by xoviat** [comment]
Just look at what @wesm said. He agrees with my central point, except perhaps not the way I phrased it (@wesm you can't support c++14 and Python 2.7 without major hacks).

**Comment by mingwandroid** [comment]
Mmm, @wesm was talking about Linux, again, this is further conflation that doesn't help your arguments. To me it seems you are grasping at whatever you can.

Clearly your goals are not those of conda-forge who aims to build provide builds of software for Python 2 and Python 3 across Windows, macOS and Linux. I would ask you to refrain then from trying to sway this discussion.

**Comment by xoviat** [comment]
And again, none of this discussion is going to change anything. What I think ultimately doesn't matter. It's what library maintainers and python developers think.

**Comment by mingwandroid** [comment]
Sure, if you convince various upstreams to stop supporting Python 2 or Python 2 on Windows then it will become too much work for us to undo your sabotage, but go to town if that's what you think you must do.

**Comment by mingwandroid** [comment]
This discussion is about conda-forge and conda-defaults unifying on the same major version of conda-build and the same compilers.

I do long for the day when flang can be used and python 2 is dead, but it's not today, and it's not next year either.

**Comment by mingwandroid** [comment]
> long for the day

.. is a bit much; better to say I look forward to it!

**Comment by xoviat** [comment]
> Mmm, @wesm was talking about Linux, again, this is further conflation that doesn't help your arguments. To me it seems you are grasping at whatever you can.

Sorry, I meant Python 2.7 on windows. My mistake.

> Clearly your goals are not those of conda-forge who aims to build provide builds of software for Python 2 and Python 3 across Windows, macOS and Linux. I would ask you to refrain then from trying to sway this discussion.

Again, what I think doesn't matter. I only care about this to the extent that I have a reasonable way to install a CRT compatible fortran compiler on windows. I'm going to head out at this point.

> Sure, if you convince various upstreams to stop supporting Python 2 or Python 2 on Windows then it will become too much work for us to undo your sabotage, but go to town if that's what you think you must do.

I mean if projects drop support for Python 2 on windows, then Python 2 is dead on windows. Right?

> .. is a bit much; better to say I look forward to it!

Then we're in agreement on this issue. It's just that I've taken concrete steps to achieve this goal.

**Comment by SylvainCorlay** [comment]
> Just look at what @wesm said. He agrees with my central point, except perhaps not the way I phrased it (@wesm you can't support c++14 and Python 2.7 without major hacks).

@xoviat I have been building C++14 extensions for Python 2.7 for a while without issue.

On conda, adding a runtime dependency to the msvc14 runtime only on windows

- vs2015_runtime # [win]

then populating the `bld.bat` with the call setting up the use of VS14 does the job.

```
if "%ARCH%" == "32" (set PLATFORM=x86) else (set PLATFORM=x64)
call "%VS140COMNTOOLS%\..\..\VC\vcvarsall.bat" %PLATFORM%
set DISTUTILS_USE_SDK=1
set MSSdk=1
"%PYTHON%" setup.py install
if errorlevel 1 exit 1
```

As said earlier by @mingwandroid, as soon as you don't deallocate memory allocated in one dll in another, (which is bad design), you should be fine with this.

**Comment by msarahan** [comment]
@xoviat it seems like you did really nice work on the scipy wheels.  Why did you use mingw there, but say it's not an option here?  I agree that long-term flang will be the best option, but like Ray, I think it'll take time.

**Comment by xoviat** [comment]
> I agree that long-term flang will be the best option, but like Ray, I think it'll take time.

+1. That's always been my claim.

> Why did you use mingw there, but say it's not an option here?

MinGW doesn't currently support reverse-symbol dependencies (although in theory it could). So you can have a fortran symbol that you need in C, but you can't have a C symbol that you need in fortran. Luckily, scipy doesn't have any C symbols that it needs in fortran, but that's not a seamless experience, which is the goal.

**Comment by xoviat** [comment]
I will say this: if Anaconda can convince the Python core developers to release a theoretical Python 2.8, then I promise not to try to get Python 2 support dropped. @SylvainCorlay is correct. I knew about that option as well (although I admit that I did not consider it until he brought it up), but that's not a seamless experience either, because you can't just developers to do the right thing.

My goal is for people to run `pip install .` in a project source directory and have the project build. Not 50% of the time. Not 80% of the time, but nearly all of the time, and for dependencies to be installed with a package manager (conda happens to be the best option on windows) just like they are on OSX with homebrew and ubuntu with apt.

**Comment by ocefpaf** [comment]
Guys, this thread became unmanageable and unproductive. There are many good points here though that we cannot miss. B/c of that I am locking this thread and we should break it down later, after the meeting, into actionable topics.

PS:  I for one would love to see a hacked 2.8 out there support modern compiler on Windows, but I don't think we, conda-forge, should provide that as the maintenance cost is beyond what we can bare.

**Comment by msarahan** [comment]
Compiler meeting results: Thursday, Nov 16 9AM Central looks to be the best time.
https://anaconda.webex.com/anaconda-en/j.php?MTID=m11b5ddad66325da22bbe58d7d1c02809

Agenda / notes at https://paper.dropbox.com/doc/Compiler-meeting-20171116-1lYaDIP93JRzzQZBlicxg?_tk=share_copylink
